source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc

```diff
@@ -92,12 +92,26 @@ void EnvoyQuicServerSession::SendGoAway(quic::QuicErrorCode error_code, const st
   }
 }
 
+void EnvoyQuicServerSession::OnCanWrite() {
+  quic::QuicServerSessionBase::OnCanWrite();
+  if (inDelayedClose() && !HasDataToWrite()) {
+    if (delayed_close_state_ == DelayedCloseState::CloseAfterFlushAndWait) {
+      ASSERT(delayed_close_timer_ != nullptr);
+      delayed_close_timer_->enableTimer(delayed_close_timeout_);
+    } else {
+      closeConnectionImmediately();
+    }
```

Comment:
If there is data to write, don't we update the delay close timer to now + delay close?

Reply:
The timer is set when close() is called. I don't think we postpone it whenever we get a chance to write, do we?

Reply:
The docs say it's timeout interval after the last write, e.g.

   // The socket will be closed immediately after the buffer is flushed _or_ if a period of
    // inactivity after the last write event greater than or equal to delayed_close_timeout_ has
    // elapsed.
    CloseAfterFlush,

and I believe the TCP stack re-arms the timer after each write
https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_impl.cc#L588

cc @AndresGuedez

Reply:
Woops, I totally missed this. I changed OnCanWrite() to re-arm timer for that case now.